designdocs/passkey_p256.md

```diff
@@ -0,0 +1,556 @@
+  // deserialize to COSE key format and check the key
+  let cose_key = CoseKey::from_slice(&payload.passkey_public_key[..]);
+  let (_, x) = cose_key
+   .params
+   .iter()
+   .find(|(l, _)| l == &Label::Int(-2));
+  let (_, y) = cose_key
+   .params
+   .iter()
+   .find(|(l, _)| l == &Label::Int(-3));
+
+  // convert COSE format to P256 verifying key
+  let encoded_point =
+   EncodedPoint::from_affine_coordinates(
+    GenericArray::from_slice(&x.clone().into_bytes()),
+    GenericArray::from_slice(&y.clone().into_bytes()),
+    false,
+   );
+  let verify_key = p256::ecdsa::VerifyingKey::from_encoded_point(&encoded_point);
+
+  let passkey_signature =
+   p256::ecdsa::DerSignature::from_bytes(&payload.passkey_signature[..]);
+
+  // extract the challenge from client_data and
+  // ensure the that the challenge is the same as the call payload
+  let client_data: serde_json::Value =
+   serde_json::from_slice(&payload.passkey_client_data_json);
+  let extracted_challenge = match client_data {
+   serde_json::Value::Object(m) => {
+    let challenge = m
+     .get(&"challenge".to_string());
+    if let serde_json::Value::String(base64_url_encoded) = challenge {
+     let decoded = base64_url::decode(&base64_url_encoded).map_err(|_| {
+      TransactionValidityError::Invalid(InvalidTransaction::Custom(9))
+     })?;
+     Ok(decoded)
+    }
+   },
+   _ => Err(TransactionValidityError::Invalid(InvalidTransaction::Custom(11))),
+  }?;
+
+  let encoded_payload = payload.passkey_call.encode();
+  ensure!(
+   encoded_payload == extracted_challenge,
+   TransactionValidityError::Invalid(InvalidTransaction::Custom(12))
+  );
+
+  // prepare signing payload which is [authenticator || sha256(client_data_json)]
+  let mut passkey_signature_payload = payload.passkey_authenticator.to_vec();
+  passkey_signature_payload.extend_from_slice(&sha2_256(&payload.passkey_client_data_json));
+
+  // finally verify the PassKey signature against the payload
+  verify_key
+   .verify(&passkey_signature_payload, &passkey_signature)
+ }
+```
+
+- Some possible optimizations
+  - **Compressed Public Key**: Currently passed Public Key is in **Cose** format, and it's between 70-73 bytes.
+      If the client can parse the Cose Public Key can extract the compressed encoded key. This Public Key can
+      get reduced to 33 bytes.
+  - **Challenge data deduplication**: Currently the challenge data is duplicated in `expected_challenge` and
+      in it's serialized format inside `passkey_client_data_json`. If the client is able to parse
+      `passkey_client_data_json` and replace `challenge` field value with empty string, then during the
+      signature check we can replace that empty string with `expected_challenge` which would allow us to
+      reduce the transaction size by around **40%**. It is important that the order of the field `passkey_client_data_json`  does not change during this operation since that would generate a different signature.
```

Comment:
```suggestion
  - **Challenge data deduplication**: Currently the challenge data is duplicated in `expected_challenge` and
      in it's serialized format inside `passkey_client_data_json`. If the client is able to parse
      `passkey_client_data_json` and replace the `challenge` field value with an empty string, then during the
      signature check we can replace that empty string with `expected_challenge` which would allow us to
      reduce the transaction size by around **40%**. It is important that the order of the field `passkey_client_data_json`  does not change during this operation since that would generate a different signature.
```